lib/kafka/producer.rb

```diff
@@ -218,6 +224,17 @@ def produce(value, key: nil, topic:, partition: nil, partition_key: nil)
         partition_key: partition_key,
       )
 
+      if buffer_size >= @max_buffer_size
+        raise BufferOverflow, "Max buffer size (#{@max_buffer_size} messages) exceeded"
+      end
+
+      if buffer_bytesize + message.bytesize >= @max_buffer_bytesize
```

Comment:
For one of the checks, you add the current `message` and not for the other. Off by one-error?